File: bridges/python/src/sdk/tools/assemblyai_audio_tool.py
import json
import time
from typing import List, Dict, Any, Optional

from ..base_tool import BaseTool
from ..toolkit_config import ToolkitConfig
from ..network import Network
from .schemas.transcription_schema import TranscriptionOutput, TranscriptionSegment


class AssemblyAIAudioTool(BaseTool):
    TOOLKIT = 'music_audio'

    def __init__(self):
        super().__init__()
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)
        self.network = Network({'base_url': 'https://api.assemblyai.com'})

    @property
    def tool_name(self) -> str:
        return 'assemblyai_audio'

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config['description']

    def transcribe_to_file(
        self,
        input_path: str,
        output_path: str,
        api_key: str,
        speaker_labels: bool = True
    ) -> str:
        """
        Transcribe audio to a file using AssemblyAI's audio transcription API via SDK Network

        Args:
            input_path: Path to the audio file to transcribe
            output_path: Path to save the JSON transcription (unified format)
            api_key: AssemblyAI API key
            speaker_labels: Enable speaker diarization (default: True)

        Returns:
            The path to the transcription file
        """
        if not api_key:
            raise Exception('AssemblyAI API key is missing')

        try:
            # Step 1: Upload the audio file
            with open(input_path, 'rb') as audio_file:
                audio_data = audio_file.read()

            upload_response = self.network.request({
                'url': '/v2/upload',
                'method': 'POST',
                'headers': {
                    'Authorization': api_key,
                    'Content-Type': 'application/octet-stream'
                },
                'data': audio_data
            })

            upload_url = upload_response['data']['upload_url']

            # Step 2: Submit transcription request
            transcription_response = self.network.request({
                'url': '/v2/transcript',
                'method': 'POST',
                'headers': {
                    'Authorization': api_key,
                    'Content-Type': 'application/json'
                },
                'data': {
                    'audio_url': upload_url,
                    'speaker_labels': speaker_labels
                },
                'use_json': True
            })

            transcript_id = transcription_response['data']['id']

            # Step 3: Poll for completion
            max_attempts = 180  # 15 minutes with 5 second intervals
            attempts = 0
            transcript_data = None

            while attempts < max_attempts:
                status_response = self.network.request({
                    'url': f'/v2/transcript/{transcript_id}',
                    'method': 'GET',
                    'headers': {
                        'Authorization': api_key
                    },
                    'use_json': True
                })

                transcript_data = status_response['data']

                if transcript_data['status'] == 'completed':
                    break
                elif transcript_data['status'] == 'error':
                    error_msg = transcript_data.get('error', 'Unknown error')
                    raise Exception(f'AssemblyAI transcription failed: {error_msg}')

                # Wait 5 seconds before polling again
                time.sleep(5)
                attempts += 1

            if attempts >= max_attempts:
                raise Exception('AssemblyAI transcription timed out')

            # Step 4: Parse and save the transcription
            parsed_output = self._parse_transcription(transcript_data)

            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(parsed_output, f, indent=2, ensure_ascii=False)

            return output_path

        except Exception as e:
            raise Exception(f'AssemblyAI transcription failed: {str(e)}')

    def _parse_transcription(
        self,
        raw_output: Dict[str, Any]
    ) -> TranscriptionOutput:
        segments: List[TranscriptionSegment] = []
        speakers_set = set()

        # Use utterances for speaker-labeled segments if available
        utterances = raw_output.get('utterances', [])
        words = raw_output.get('words', [])

        if utterances and len(utterances) > 0:
            for utterance in utterances:
                speaker = utterance.get('speaker')
                segments.append({
                    'from': float(utterance.get('start', 0)) / 1000.0,  # Convert ms to seconds
                    'to': float(utterance.get('end', 0)) / 1000.0,
                    'text': utterance.get('text', ''),
                    'speaker': speaker
                })
                if speaker:
                    speakers_set.add(speaker)
        elif words and len(words) > 0:
            # Fallback to word-level data if utterances are not available
            # Group consecutive words by speaker (if available)
            current_segment = None

            for word in words:
                speaker = word.get('speaker', None)
                word_start = float(word.get('start', 0)) / 1000.0
                word_end = float(word.get('end', 0)) / 1000.0
                word_text = word.get('text', '')

                if (
                    current_segment
                    and current_segment['speaker'] == speaker
                    and word_start - current_segment['to'] < 1.0  # Max 1 second gap
                ):
                    # Extend current segment
                    current_segment['to'] = word_end
                    current_segment['text'] += f' {word_text}'
                else:
                    # Start a new segment
                    if current_segment:
                        segments.append(current_segment)
                    current_segment = {
                        'from': word_start,
                        'to': word_end,
                        'text': word_text,
                        'speaker': speaker
                    }

                if speaker:
                    speakers_set.add(speaker)

            # Push the last segment
            if current_segment:
                segments.append(current_segment)
        else:
            # Fallback: create a single segment with the full text
            audio_duration = raw_output.get('audio_duration', 0)
            segments.append({
                'from': 0.0,
                'to': audio_duration if audio_duration else 0.0,
                'text': raw_output.get('text', ''),
                'speaker': None
            })

        # Calculate duration
        audio_duration = raw_output.get('audio_duration')
        if audio_duration:
            duration = float(audio_duration) / 1000.0
        elif len(segments) > 0:
            duration = segments[-1]['to']
        else:
            duration = 0.0

        return {
            'duration': duration,
            'speakers': list(speakers_set),
            'speaker_count': len(speakers_set),
            'segments': segments,
            'metadata': {
                'tool': self.tool_name
            }
        }

File: bridges/nodejs/src/sdk/tools/assemblyai_audio-tool.ts
import fs from 'node:fs'

import type { TranscriptionOutput } from '@sdk/tools/schemas'
import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'
import { Network } from '@sdk/network'

interface AssemblyAIUploadResponse {
  upload_url: string
}

interface AssemblyAITranscriptionResponse {
  id: string
  status: 'queued' | 'processing' | 'completed' | 'error'
  text: string
  words?: {
    text: string
    start: number
    end: number
    confidence: number
    speaker?: string
  }[]
  utterances?: {
    text: string
    start: number
    end: number
    confidence: number
    speaker: string
    words: {
      text: string
      start: number
      end: number
      confidence: number
    }[]
  }[]
  audio_duration?: number
  error?: string
}

export default class AssemblyAIAudioTool extends Tool {
  private static readonly TOOLKIT = 'music_audio'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    this.config = ToolkitConfig.load(AssemblyAIAudioTool.TOOLKIT, this.toolName)
  }

  get toolName(): string {
    return 'assemblyai_audio'
  }

  get toolkit(): string {
    return AssemblyAIAudioTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Transcribe audio to a file using AssemblyAI's audio transcription API via SDK Network
   * @param inputPath Path to the audio file to transcribe
   * @param outputPath Path to save the JSON transcription
   * @param apiKey AssemblyAI API key
   * @param speakerLabels Enable speaker diarization (default: true)
   */
  async transcribeToFile(
    inputPath: string,
    outputPath: string,
    apiKey: string,
    speakerLabels = true
  ): Promise<string> {
    if (!apiKey) {
      throw new Error('AssemblyAI API key is missing')
    }

    const network = new Network({ baseURL: 'https://api.assemblyai.com' })

    // Step 1: Upload the audio file
    const audioData = await fs.promises.readFile(inputPath)
    const uploadResponse = await network.request({
      url: '/v2/upload',
      method: 'POST',
      data: audioData,
      headers: {
        Authorization: apiKey,
        'Content-Type': 'application/octet-stream'
      }
    })

    const uploadUrl = (uploadResponse.data as AssemblyAIUploadResponse)
      .upload_url

    // Step 2: Submit transcription request
    const transcriptionResponse = await network.request({
      url: '/v2/transcript',
      method: 'POST',
      data: {
        audio_url: uploadUrl,
        speaker_labels: speakerLabels
      },
      headers: {
        Authorization: apiKey,
        'Content-Type': 'application/json'
      }
    })

    const transcriptId = (
      transcriptionResponse.data as AssemblyAITranscriptionResponse
    ).id

    // Step 3: Poll for completion
    let transcriptData: AssemblyAITranscriptionResponse
    let attempts = 0
    const maxAttempts = 180 // 15 minutes with 5 second intervals

    while (attempts < maxAttempts) {
      const statusResponse = await network.request({
        url: `/v2/transcript/${transcriptId}`,
        method: 'GET',
        headers: {
          Authorization: apiKey
        }
      })

      transcriptData = statusResponse.data as AssemblyAITranscriptionResponse

      if (transcriptData.status === 'completed') {
        break
      } else if (transcriptData.status === 'error') {
        throw new Error(
          `AssemblyAI transcription failed: ${
            transcriptData.error || 'Unknown error'
          }`
        )
      }

      // Wait 5 seconds before polling again
      await new Promise((resolve) => setTimeout(resolve, 5000))
      attempts++
    }

    if (attempts >= maxAttempts) {
      throw new Error('AssemblyAI transcription timed out')
    }

    // Step 4: Parse and save the transcription
    const parsedOutput = this.parseTranscription(transcriptData!)

    await fs.promises.writeFile(
      outputPath,
      JSON.stringify(parsedOutput, null, 2),
      'utf8'
    )

    return outputPath
  }

  private parseTranscription(
    rawOutput: AssemblyAITranscriptionResponse
  ): TranscriptionOutput {
    const segments: {
      from: number
      to: number
      text: string
      speaker: string | null
    }[] = []
    const speakers: Set<string> = new Set()

    // Use utterances for speaker-labeled segments if available
    if (rawOutput.utterances && rawOutput.utterances.length > 0) {
      for (const utterance of rawOutput.utterances) {
        segments.push({
          from: utterance.start / 1_000, // Convert milliseconds to seconds
          to: utterance.end / 1_000,
          text: utterance.text,
          speaker: utterance.speaker
        })
        speakers.add(utterance.speaker)
      }
    } else if (rawOutput.words && rawOutput.words.length > 0) {
      // Fallback to word-level data if utterances are not available
      // Group consecutive words by speaker (if available)
      let currentSegment: {
        from: number
        to: number
        text: string
        speaker: string | null
      } | null = null

      for (const word of rawOutput.words) {
        const speaker = word.speaker || null

        if (
          currentSegment &&
          currentSegment.speaker === speaker &&
          word.start / 1_000 - currentSegment.to < 1.0 // Max 1 second gap
        ) {
          // Extend current segment
          currentSegment.to = word.end / 1_000
          currentSegment.text += ` ${word.text}`
        } else {
          // Start a new segment
          if (currentSegment) {
            segments.push(currentSegment)
          }
          currentSegment = {
            from: word.start / 1_000,
            to: word.end / 1_000,
            text: word.text,
            speaker: speaker
          }
        }

        if (speaker) {
          speakers.add(speaker)
        }
      }

      // Push the last segment
      if (currentSegment) {
        segments.push(currentSegment)
      }
    } else {
      // Fallback: create a single segment with the full text
      segments.push({
        from: 0,
        to: (rawOutput.audio_duration || 0) / 1_000,
        text: rawOutput.text,
        speaker: null
      })
    }

    // Calculate duration
    let duration = rawOutput.audio_duration ? rawOutput.audio_duration : 0
    if (!duration && segments.length > 0) {
      duration = segments[segments.length - 1]?.to || 0
    }

    return {
      duration,
      speakers: Array.from(speakers),
      speaker_count: speakers.size,
      segments,
      metadata: {
        tool: this.toolName
      }
    }
  }
}
